README.md

```diff
@@ -7,5 +7,164 @@
+[`dogma.Application`] interface.
+
+### Engine
+
+An **engine** is the platform upon which an [application](#application) is
+executed. The engine is responsible for the delivery of messages and the
+persistence of application data.
+
+This module does not provide an engine implementation, but rather defines the
+API that sits between the application and the engine. The API documentation for
+each interface indicates whether the implementation is to be provided by the
+engine or the application.
+
+### Aggregate
+
+An **aggregate** is a unit of application logic with associated state that
+encodes the business invariants of a specific application. These invariants are
+the "rules" of the business domain that must not be violated, even temporarily.
+This concept is taken directly from [Domain Driven Design].
+
+An aggregate receives command [messages](#message-and-message-handler) in order
+to effect a change in a particular **instance** of that aggregate. Such state
+changes are represented by event messages. By definition, changes to the state
+of an aggregate instance are ["immediately consistent"][Immediate Consistency]
+(aka "transactionally consistent"). This means that the results of a command
+against a given instance are always visible to subsequent commands for that
+instance.
+
+Aggregate state is managed by the [engine](#engine), ensuring that state
+transitions occur atomically with the recording of the events that describe
+those state
+transitions.
+
+Aggregates can be quite a difficult concept to grasp. The book [Domain Driven
+Design Distilled], by Vaugn Vernon offers a suitable introduction to aggregates
+and the other elements of domain driven design.
+
+Aggregates are represented by the [`dogma.AggregateMessageHandler`] interface.
+
+### Process
+
+A **process** is a unit of application logic with associated state that serves
+to automate some long running business process. In particular, they can be used
+to coordinate changes across multiple [aggregate](#aggregate) instances, or
+between aggregates and [integrations](#integration).
+
+Processes receive event [messages](#message-and-message-handler) and produce
+command messages. Like aggregates, the received events are routed to a specific
+instance.
+
+Additionally, processes can produce timeout messages, which are routed back to
+the same process instance at a specific time. Such messages are used to
+implement processes that incorporate some time-based component.
+
+Because a process coordinates changes within the application using multiple
+messages, and each message represents a single atomic change to the application's
+state, the changes made by a process are ["eventually consistent"][Eventual Consistency].
+The precise guarantees regarding process consistency are specific to the [engine](#Engine) implementation.
+
+Process state is managed by the engine, ensuring that state transitions occur
+atomically with the enqueing of the commands the processes produce.
```

Comment:
```suggestion
atomically with the enqueing of the commands that the processes produce.
```

Reply:
I propose rewording this whole paragraph:

```
Aggregate state is managed by the [engine](#engine), ensuring that changes to a
specific instance and the recording of events that represent those changes occur
atomically.
```

Reply:
that sounds clear to me